rfd/0163-vnet.md

```diff
@@ -0,0 +1,792 @@
+API](https://forums.developer.apple.com/forums/thread/701118?answerId=705661022#705661022), it gets
+us past the problem of acquiring root privileges and lets us focus on validating VNet as a tool.
+In the [Alternatives to osascript](#alternatives-to-osascript) section, we describe other approaches
+which offer better UX at the cost of development time.
+
+#### Guiding the user towards VNet
+
+Before VNet, the user would access web and TCP apps through the unified resources view in Connect:
+
+<img width="871" alt="Unified Resources view in Connect v15.0.0" src="assets/0163-apps-in-connect-v15.png">
+
+Clicking "Launch" next to a web app would launch this app in the browser through the `/web/launch`
+URL of the proxy service. The user could also set up a local proxy for the specific app by clicking
+on the three dots menu and choosing "Set up connection". For TCP apps, clicking "Connect" would set
+up a local proxy for that app.
+
+This changes with the introduction of VNet. Here's a screenshot from Teleport Connect in cluster
+`teleport-local.dev` running VNet:
+
+<img width="871" alt="Unified Resources view in Connect with VNet turned on" src="assets/0163-apps-with-vnet.png">
+
+Web apps now have two actions available in the three dots menu: "Connect and copy VNet address" and
+"Connect to local port". "Connect to local port" starts a regular local proxy. This is kept for
+users who might not be able to use VNet, either because they lack root privileges or VNet would
+conflict with other software running on the computer.
+
+"Connect and copy VNet address" opens the VNet tab and shows the root privileges prompt. While the
+prompt is shown, the VNet tab is in the "starting" state, explaining what VNet is.
+
+<img width="871" alt="VNet starting" src="assets/0163-vnet-starting.png">
+
+Once VNet starts, the VNet address of the app is copied to clipboard. Afterwards the button becomes
+just "Copy VNet address" when VNet is running. Copying the address copies `spec.public_addr`.
+
+For TCP apps, the main "Connect" button is the equivalent of "Connect and copy VNet address". From
+the three dots menu, the user can select the sole option "Connect to local port". The address
+displayed on the card of a TCP app is the `spec.public_addr` of the app.
+
+Selecting an app from the search bar performs the same action as clicking on the main button in the
+app card.
+
+#### Managing VNet
+
+<img width="871" alt="VNet in Connections" src="assets/0163-vnet-in-connections.png">
+
+VNet has a persistent place in the UI in the connection list in top left. From there, the user can
+see VNet status for the current root cluster and quickly stop VNet for all root clusters, similar to
+how they can disconnect existing database, SSH and Kube connections.
+
+Clicking on the VNet item opens a dedicated VNet tab. Data in this tab is scoped to the root cluster
+of the current workspace. The user can see the status of VNet for the root cluster, along with an
```

Comment:
Could we simplify by dropping the cluster scopes and making it global? Or are the tabs typically scoped by cluster

Reply:
Yeah, that was the problem, tabs within a workspace are always scoped by root cluster. I also thought that keeping VNet off for certain clusters might be useful if there are conflicts and whatnot. But I think it's safe to safe it was a premature solution to a problem that might not even exist in the first place.

I spent today on moving the UI to the connections list – that's one element of the UI that's always been "global". There are still a few things I need to polish, but I like it much better. It also offers some potential for further integration of VNet with Connect, e.g. the connections from VNet could start showing up in the regular connections list.

Clicking on an app in the list copies the address to the clipboard (together with an error if there's any, but you can also hover over it to see the full message).

I assume people won't be using this panel very often anyway, it's for debugging purposes mostly. I'll update the RFD tomorrow morning.

https://github.com/gravitational/teleport/assets/27113/14839408-8471-4f4f-8d13-06f30e3e0f77